design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql, Map<String, Object> properties);
+```
+where `QueryResult` is as follows:
+```
+/**
+ * The result of a query (push or pull), streamed one row at time. Records may be consumed by either
+ * subscribing to the publisher or polling (blocking) for one record at a time. These two methods of
+ * consumption are mutually exclusive; only one method may be used (per QueryResult).
+ */
+public interface QueryResult extends Publisher<Row> {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  String queryID();
+
+  /**
+   * Block until a row becomes available.
+   *
+   * @return the row.
+   */
+  Row poll();
+
+  /**
+   * Block until a row becomes available or the timeout has elapsed.
+   *
+   * @param timeout amount of to wait for a row. Non-positive values are interpreted as no timeout.
+   * @param timeUnit unit for timeout param.
+   * @return the row, if available; else, null.
+   */
+  Row poll(long timeout, TimeUnit timeUnit);
+
+  boolean isComplete();
+
+  void close();
+}
+```
+Note that `QueryResult` is a Reactive Streams `Publisher` so users can stream results. Users can also call `poll()` to receive
+results in a synchronous fashion instead. Only one of the two methods will be allowed per `QueryResult` instance.
+
+The `Row` interface is as follows:
+```
+/**
+ * A single record, returned as part of a query result.
+ */
+public interface Row {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  List<Object> values();
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Object getObject(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Object getObject(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  String getString(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  String getString(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Integer getInt(int columnIndex);
```

Comment:
I assume we return `Integer` instead if `int` to handle `NULL` ?

Would it make sense to add `isNull(int columnIndex)` and `isNotNull(int columnIndex)`, too? (also for `columnName`)

Reply:
+1 isNull would be useful too

Reply:
I don't think isNotNull() really add much.

Reply:
Was just an idea -- and if there is demand coming up,`isNotNull()` can always be added later.